In [1]:
import pandas as pd
import numpy as np
import json

In [2]:
df_real = pd.read_csv('../Cleaned Data/Real_News.csv')
df_real = df_real[['url','clean_title','clean_text']]
print(df_real.shape)
df_real.head()



(905, 3)


url  \
0  https://www.cnn.com/2020/03/29/politics/trump-...   
1  https://www.cnn.com/world/live-news/coronaviru...   
2  https://www.cnn.com/2020/03/29/politics/trump-...   
3  https://www.cnn.com/2020/03/29/opinions/donald...   
4  https://www.cnn.com/2020/03/29/politics/us-ele...   

                                         clean_title  \
0  trump extends federal social distancing guidel...   
1     coronavirus live updates global pandemic kills   
2     trump concedes us coronavirus death toll could   
3                          donald trumps special day   
4                        long trumps poll surge last   

                                          clean_text  
0  washington cnn president donald trump said sun...  
1  exterior elmhurst hospital queens sunday march...  
2  cnn president donald trump acknowledged sunday...  
3  sign get new weekly column newsletter looking ...  
4  cnn  race publicly least largely halted countr...

In [3]:
df_real['label'] = 1
df_real.dropna(inplace=True)
print(df_real.shape)
df_real.drop(['url'],axis=1,inplace=True)
df_real.head()

(905, 4)


clean_title  \
0  trump extends federal social distancing guidel...   
1     coronavirus live updates global pandemic kills   
2     trump concedes us coronavirus death toll could   
3                          donald trumps special day   
4                        long trumps poll surge last   

                                          clean_text  label  
0  washington cnn president donald trump said sun...      1  
1  exterior elmhurst hospital queens sunday march...      1  
2  cnn president donald trump acknowledged sunday...      1  
3  sign get new weekly column newsletter looking ...      1  
4  cnn  race publicly least largely halted countr...      1

In [4]:
df_fake = pd.read_csv('../Cleaned Data/Human_Fake_News.csv')

In [5]:
print(df_fake.shape)
df_fake.head()

(12273, 23)


Unnamed: 0                                      uuid  ord_in_thread  \
0           0  6a175f46bcd24d39b3e962ad0f29936721db70db              0   
1           1  2bdc29d12605ef9cf3f09f9875040a7113be5d5b              0   
2           2  c70e149fdd53de5e61c29281100b9de0ed268bc3              0   
3           3  7cf7c15731ac2a116dd7f629bd57ea468ed70284              0   
4           4  0206b54719c7e241ffe0ad4315b808290dbe6c0f              0   

                 author                      published  \
0     Barracuda Brigade  2016-10-26T21:41:00.000+03:00   
1  reasoning with facts  2016-10-29T08:47:11.259+03:00   
2     Barracuda Brigade  2016-10-31T01:41:49.479+02:00   
3                Fed Up  2016-11-01T05:22:00.000+02:00   
4                Fed Up  2016-11-01T21:56:00.000+02:00   

                                               title  \
0  Muslims BUSTED: They Stole Millions In Gov’t B...   
1  Re: Why Did Attorney General Loretta Lynch Ple...   
2  BREAKING: Weiner Cooperating With FBI On Hilla...   
3  PIN DROP SPEECH BY FATHER OF DAUGHTER Kidnappe...   
4  FANTASTIC! TRUMP'S 7 POINT PLAN To Reform Heal...   

                                                text language  \
0  Print They should pay all the back all the mon...  english   
1  Why Did Attorney General Loretta Lynch Plead T...  english   
2  Red State : \nFox News Sunday reported this mo...  english   
3  Email Kayla Mueller was a prisoner and torture...  english   
4  Email HEALTHCARE REFORM TO MAKE AMERICA GREAT ...  english   

                         crawled             site_url  ... spam_score  \
0  2016-10-27T01:49:27.168+03:00  100percentfedup.com  ...      0.000   
1  2016-10-29T08:47:11.259+03:00  100percentfedup.com  ...      0.000   
2  2016-10-31T01:41:49.479+02:00  100percentfedup.com  ...      0.000   
3  2016-11-01T15:46:26.304+02:00  100percentfedup.com  ...      0.068   
4  2016-11-01T23:59:42.266+02:00  100percentfedup.com  ...      0.865   

                                        main_img_url replies_count  \
0  http://bb4sp.com/wp-content/uploads/2016/10/Fu...             0   
1  http://bb4sp.com/wp-content/uploads/2016/10/Fu...             0   
2  http://bb4sp.com/wp-content/uploads/2016/10/Fu...             0   
3  http://100percentfedup.com/wp-content/uploads/...             0   
4  http://100percentfedup.com/wp-content/uploads/...             0   

   participants_count likes  comments  shares  type  \
0                   1     0         0       0  bias   
1                   1     0         0       0  bias   
2                   1     0         0       0  bias   
3                   0     0         0       0  bias   
4                   0     0         0       0  bias   

                                          clean_text  \
0  print pay back money plus interest entire fami...   
1  attorney general loretta lynch plead fifth bar...   
2  red state  fox news sunday reported morning an...   
3  email kayla mueller prisoner tortured isis cha...   
4  email healthcare reform make america great sin...   

                                         clean_title  
0        muslims busted stole millions govt benefits  
1         attorney general loretta lynch plead fifth  
2  breaking weiner cooperating fbi hillary email ...  
3  pin drop speech father daughter kidnapped kill...  
4  fantastic! trumps  point plan reform healthcar...  

[5 rows x 23 columns]

In [6]:
df_fake = df_fake[['site_url','clean_title','clean_text']]
df_fake.dropna(inplace=True)
df_fake.rename(columns={"site_url": "url"}, inplace=True)
chosen_idx = np.random.choice(df_fake.shape[0], replace=False, size=df_real.shape[0])
df_fake_trim = df_fake.iloc[chosen_idx]

In [7]:
df_fake_trim['label'] = 0
print(df_fake_trim.shape)
df_fake_trim.drop(['url'], axis=1, inplace= True)
df_fake_trim.head()


(905, 4)


/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4102: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


clean_title  \
10903  mystery solved! gets hillary clinton get movin...   
9087   rigged election hillary  trump caught partying...   
7026                         nato put war footing russia   
10636  madman cathedral former monk dedicates yrs bui...   
7480   hillary cancels public appearance due large cr...   

                                              clean_text  label  
10903  liars never win liarsneverwin october   hard t...      0  
9087   covert geopolitics suspicious donald trump sin...      0  
7026   nato put war footing russia  jens stoltenberg ...      0  
10636  check keiser report website httpwwwmaxkeiserco...      0  
7480   memes breaking hillary cancels public appearan...      0

In [8]:
df_grover = pd.read_csv('../Cleaned Data/Neural_Fake_News.csv')
df_grover = df_grover[['clean_title','clean_text']]
print(df_grover.shape)
df_grover.dropna(inplace=True)
print(df_grover.shape)
df_grover.head()
df_grover['label'] = 2
df_grover.head()

(908, 2)
(908, 2)


clean_title  \
0  trump extends federal social distancing guidel...   
1     coronavirus live updates global pandemic kills   
2     trump concedes us coronavirus death toll could   
3                            tracking covid cases us   
4                          donald trumps special day   

                                          clean_text  label  
0  president donald trump first official state vi...      2  
1  dylan welch april   bloomberg  outbreak intest...      2  
2  donald trump home state california sunday land...      2  
3  editors note readers visitors new york area ma...      2  
4  wish many press morning president facebook sta...      2

# Experiment Number 1
## Naive Bayes, Real Human News and Fake Human News
905 instances of each 

Class: Label

Real: 0

Fake: 1

In [9]:
#combine both the real and trimmed fake news into one df
combined = pd.concat([df_real,df_fake_trim],ignore_index=True)

In [10]:
print(combined.head())
combined.tail()

                                         clean_title  \
0  trump extends federal social distancing guidel...   
1     coronavirus live updates global pandemic kills   
2     trump concedes us coronavirus death toll could   
3                          donald trumps special day   
4                        long trumps poll surge last   

                                          clean_text  label  
0  washington cnn president donald trump said sun...      1  
1  exterior elmhurst hospital queens sunday march...      1  
2  cnn president donald trump acknowledged sunday...      1  
3  sign get new weekly column newsletter looking ...      1  
4  cnn  race publicly least largely halted countr...      1  


clean_title  \
1805  madonna says shell give blowjob anyone votes h...   
1806  never want use johnson  johnson baby powder re...   
1807  new gallup poll reveals americans really think...   
1808                                       lefts vision   
1809    open letter black south african police officers   

                                             clean_text  label  
1805  home  news  madonna says shell give blowjob an...      0  
1806  jury st louis awarded california woman  millio...      0  
1807  home  us  new gallup poll reveals americans re...      0  
1808  thomas sowell november   political left keeps ...      0  
1809  tweet widget black power front students joinin...      0

In [11]:
import nltk
from wordcloud import WordCloud
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score, accuracy_score, recall_score, precision_score

In [12]:
combined_text = combined['clean_text'].values
X_title =combined['clean_title'].values
labels = combined['label'].values
combined.groupby('label').agg('count')

clean_title  clean_text
label                         
0              905         905
1              905         905

In [13]:
tfidf = TfidfVectorizer(ngram_range=(1,2), max_df= 0.85, min_df= 0.01)

In [14]:
combined_text = tfidf.fit_transform(combined_text)
# combined['title'] = tfidf.fit_transform (combined['title'])

In [15]:
print("tf idf vectorized text shape: {}".format(combined_text.shape))
print("original text shape: {}".format(combined['clean_text'].shape))


<class 'scipy.sparse.csr.csr_matrix'>
tf idf vectorized text shape: (1810, 5322)
original text shape: (1810,)


In [16]:
train_text, test_text, train_labels, test_labels  = train_test_split(combined_text, labels, test_size=0.2)


In [44]:
import collections

print("size of train_text: {}".format(train_text.shape))

print("size of train_labels: {}".format(train_labels.shape))
collections.Counter(train_labels)


<class 'scipy.sparse.csr.csr_matrix'>
size of train_text: (1448, 5092)
<class 'numpy.ndarray'>
size of train_labels: (1448,)


Counter({1: 724, 0: 724})

In [45]:
print("size of test_text: {}".format(test_text.shape))

print("size of test_labels: {}".format(test_labels.shape))
collections.Counter(test_labels)



<class 'scipy.sparse.csr.csr_matrix'>
size of test_text: (362, 5092)
<class 'numpy.ndarray'>
size of test_labels: (362,)


Counter({1: 181, 0: 181})

In [19]:
from sklearn.naive_bayes import MultinomialNB
nb = MultinomialNB()


In [20]:
nb.fit(train_text,train_labels)

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [21]:
train_pred = nb.predict(train_text)

In [22]:
print('Naive Bayes In Sample F1 and Accuracy Scores:')
print('F1 score {:.4}%'.format(f1_score(train_labels, train_pred, average='macro')*100 ))
print ('Accuracy score {:.4}%'.format(accuracy_score(train_labels, train_pred)*100))

Naive Bayes In Sample F1 and Accuracy Scores:
F1 score 96.89%
Accuracy score 96.89%


In [23]:
combined.loc[np.where(train_labels != train_pred)[0].tolist()].groupby('label').agg('count')

clean_title  clean_text
label                         
0               16          16
1               29          29

In [24]:
test_pred = nb.predict(test_text)
print('Naive Bayes Out of Sample F1 and Accuracy Scores:')
print('F1 score {:.4}%'.format(f1_score(test_labels, test_pred, average='macro')*100 ))
print ('Accuracy score {:.4}%'.format(accuracy_score(test_labels, test_pred)*100))

Naive Bayes Out of Sample F1 and Accuracy Scores:
F1 score 93.63%
Accuracy score 93.65%


In [25]:
combined.loc[np.where(test_labels != test_pred)[0].tolist()].groupby('label').agg('count')

clean_title  clean_text
label                         
1               23          23

# Experiment 2
## Augmenting with Neural Fake News

905 Real News, 905 Fake (neural and human combined)

80% Train = 1448
20% Test = 362 (1/2 Real, 1/2 Human Fake)

Train on (1/2 real news, 1/2 fake news (1/4 human fake + 1/4 neural fake))

Human Fake = 543 (362 for train, 181 for test)

Neural Fake = 362 for train

In [26]:
print(df_grover.shape)
print(df_real.shape)
print(df_fake.shape)

(908, 3)
(905, 3)
(12160, 3)


In [27]:
trim_size = int(df_real.shape[0] * 0.6)
chosen_idx = np.random.choice(df_fake.shape[0], replace=False, size=trim_size)
df_fake_trim_2 = df_fake.iloc[chosen_idx]
df_fake_trim_2['label'] = 0
print(df_fake_trim_2.shape)
df_fake_trim_2.drop(['url'], axis=1, inplace= True)
df_fake_trim_2.head()

(543, 4)


/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.
/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4102: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


clean_title  \
4737   politics populism mudslinging undermines democ...   
11030                                     oh lovely war!   
6353   mayhem awaits crooked hillary clinton advised ...   
9477   time! cnn fires donna brazile rigging debates ...   
655    turkeys rivalry kurds likely complicate raqqa ...   

                                              clean_text  label  
4737   politics populism mudslinging undermines democ...      0  
11030  oh lovely war! delusional foreign policy could...      0  
6353   mayhem awaits crooked hillary clinton advised ...      0  
9477   email cnn made official monday firing commenta...      0  
655    turkey repeatedly demanded us allow kurds part...      0

In [28]:
trim_size = int(df_real.shape[0]*0.4)
chosen_idx = np.random.choice(df_grover.shape[0], replace=False, size=trim_size)
df_grover_trim = df_grover.iloc[chosen_idx]
df_grover_trim['label'] = 0
print(df_grover_trim.shape)
df_grover_trim.head()

(362, 3)


/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


clean_title  \
313  customer calls companies rise amid coronavirus...   
193  benny gantz breaking vows says would serve net...   
164                        shielding fetus coronavirus   
212      rhode island pulls new yorkers keep virus bay   
205                       much public know coronavirus   

                                            clean_text  label  
313  warning signs surfaced market wireless communi...      0  
193  rep bart custer dohio calling upcoming  budget...      0  
164  image pbs humans microbes vulnerable various v...      0  
212  rhode islands attorney general ashley schmidt ...      0  
205  please enable javascript watch video hollywood...      0

In [29]:
print(df_real.shape)
df_real.head()

(905, 3)


clean_title  \
0  trump extends federal social distancing guidel...   
1     coronavirus live updates global pandemic kills   
2     trump concedes us coronavirus death toll could   
3                          donald trumps special day   
4                        long trumps poll surge last   

                                          clean_text  label  
0  washington cnn president donald trump said sun...      1  
1  exterior elmhurst hospital queens sunday march...      1  
2  cnn president donald trump acknowledged sunday...      1  
3  sign get new weekly column newsletter looking ...      1  
4  cnn  race publicly least largely halted countr...      1

In [30]:
#Create Test Set first from trimmed
trim_size = int(df_real.shape[0]*.2)
mask =  np.random.choice(df_real.shape[0], trim_size, replace=False) 
neg_mask = [i for i in range(0, df_real.shape[0]) if i not in mask]
df_real_test = df_real.iloc[mask]
df_real_train = df_real.iloc[neg_mask]
print(df_real_test.shape)
print(df_real_train.shape)

(181, 3)
(724, 3)


In [31]:
#Create Test Set first from trimmed
print(df_fake_trim_2.shape)
trim_size = int(df_fake_trim_2.shape[0]*(1/3))
mask =  np.random.choice(df_fake_trim_2.shape[0], trim_size, replace=False) 
neg_mask = [i for i in range(0, df_fake_trim_2.shape[0]) if i not in mask]
df_fake_trim_2_test = df_fake_trim_2.iloc[mask]
df_fake_trim_2_train = df_fake_trim_2.iloc[neg_mask]
print(df_fake_trim_2_test.shape)
print(df_fake_trim_2_train.shape)

(543, 3)
(181, 3)
(362, 3)


In [32]:
#combine both the real and trimmed fake news into one df
combined_train = pd.concat([df_real_train,df_fake_trim_2_train,df_grover_trim],ignore_index=True)
combined_test = pd.concat([df_real_test,df_fake_trim_2_test],ignore_index=True)

In [33]:
print(combined_train.shape)
print(combined_train.head())
combined_train.tail()

(1448, 3)
                                         clean_title  \
0  trump extends federal social distancing guidel...   
1     coronavirus live updates global pandemic kills   
2                          donald trumps special day   
3                        long trumps poll surge last   
4  breakdown false misleading statements trump ro...   

                                          clean_text  label  
0  washington cnn president donald trump said sun...      1  
1  exterior elmhurst hospital queens sunday march...      1  
2  sign get new weekly column newsletter looking ...      1  
3  cnn  race publicly least largely halted countr...      1  
4  cnn president donald trump made numerous false...      1  


clean_title  \
1443      irs coronavirusextended deadlines apply taxes   
1444        tornado rips arkansas city injuring  people   
1445     online festival celebrates arts affected covid   
1446  happy little miracle dark times plant nursery ...   
1447   dr deborah birx urging calm come heavy criticism   

                                             clean_text  label  
1443  reuters  office personnel management evaluatin...      0  
1444  kansas city mo tornado rips arkansas city caus...      0  
1445  vancouver prince edward island announced  inau...      0  
1446  whole plants often make life impossible small ...      0  
1447  august  ad age small business risk  insurance ...      0

In [34]:
print(combined_test.shape)
print(combined_test.head())
combined_test.tail()

(362, 3)
                                         clean_title  \
0          life time coronavirus photos around world   
1  pentagon asks reconsider massive cloud contrac...   
2  experts converge plans easing coronavirus rest...   
3                    democratic candidates president   
4  california dmv closing field offices public st...   

                                          clean_text  label  
0  newsletter mustread stories la times get days ...      1  
1  pentagon wishes reconsider award controversial...      1  
2  proposals come white houses current day guidel...      1  
3  race democratic presidential nomination two ca...      1  
4  faced concerns among workers spread covid cali...      1  


clean_title  \
357  breaking jury finds #oregonstandoff seven incl...   
358      turkeys face deep divisions thanksgiving vote   
359  cnn over! look trump moments ago end wolf blit...   
360    king tut eastern european according dna results   
361  #trumpprotest communists mobilize disrupt pres...   

                                            clean_text  label  
357  breaking jury finds #oregonstandoff seven incl...      0  
358  november   group  white holland turkeys gigant...      0  
359  story paris swade  cnn cnn over! need black te...      0  
360  posted eddie scientists zurichbased dna geneal...      0  
361  home  articles bound  #trumpprotest communists...      0

In [48]:
train_text = combined_train['clean_text'].values
test_text = combined_test['clean_text'].values
train_labels = combined_train['label'].values
test_labels = combined_test['label'].values

In [36]:
type(train_text)
print(train_text.shape)
print(test_text.shape)
fit_text = np.concatenate((train_text, test_text))
print(fit_text.shape)

(1448,)
(362,)
(1810,)


In [37]:
tfidf_2 = TfidfVectorizer(ngram_range=(1,2), max_df= 0.85, min_df= 0.01)
tfidf_2.fit(fit_text)
train_text = tfidf_2.transform(train_text)
print(type(train_text))
print("tf idf vectorized text shape: {}".format(train_text.shape))
print("original text shape: {}".format(combined_train['clean_text'].shape))

test_text = tfidf_2.transform(test_text)
print(type(test_text))
print("tf idf vectorized text shape: {}".format(test_text.shape))
print("original text shape: {}".format(combined_test['clean_text'].shape))

<class 'scipy.sparse.csr.csr_matrix'>
tf idf vectorized text shape: (1448, 5092)
original text shape: (1448,)
<class 'scipy.sparse.csr.csr_matrix'>
tf idf vectorized text shape: (362, 5092)
original text shape: (362,)


In [38]:
nb_2 = MultinomialNB()
nb_2.fit(train_text,train_labels)
train_pred = nb_2.predict(train_text)

In [39]:
print('Naive Bayes In Sample F1 and Accuracy Scores:')
print('F1 score {:.4}%'.format(f1_score(train_labels, train_pred, average='macro')*100 ))
print ('Accuracy score {:.4}%'.format(accuracy_score(train_labels, train_pred)*100))

Naive Bayes In Sample F1 and Accuracy Scores:
F1 score 91.37%
Accuracy score 91.37%


In [40]:
combined.loc[np.where(train_labels != train_pred)[0].tolist()].groupby('label').agg('count')

clean_title  clean_text
label                         
0               61          61
1               64          64

In [41]:
test_pred = nb_2.predict(test_text)
print('Naive Bayes Out of Sample F1 and Accuracy Scores:')
print('F1 score {:.4}%'.format(f1_score(test_labels, test_pred, average='macro')*100 ))
print ('Accuracy score {:.4}%'.format(accuracy_score(test_labels, test_pred)*100))

Naive Bayes Out of Sample F1 and Accuracy Scores:
F1 score 93.07%
Accuracy score 93.09%


In [42]:
combined.loc[np.where(test_labels != test_pred)[0].tolist()].groupby('label').agg('count')

clean_title  clean_text
label                         
1               25          25